content/advent-2018/avoid-gc-overhead-large-heaps.md

```diff
@@ -0,0 +1,376 @@
+	}
+}
+```
+
+```
+a[0] is C00009A000
+*a[0] is 0
+a[1] is C00009A040
+*a[1] is 1
+a[2] is C00009A050
+*a[2] is 2
+
+*a[0] is 0
+*a[1] is 1
+*a[2] is 2
+```
+## The nub of the problem
+
+So, it turns out that pointers are the enemy, both when we have large amounts of memory allocated on-heap, and when we try to work around this by moving the data to our own off-heap allocations. If we can avoid any pointers within the types we're allocating they won't cause GC overhead, so we won't need to use any off-heap tricks. If we do use off-heap allocations, then we need to avoid storing pointers to heap allocations unless these are also referenced by memory that is visible to the GC.
+
+# How can we avoid pointers?
+
+In large heaps, pointers are evil and must be avoided. But you need to be able to spot them to avoid them, and they aren't always obvious. Strings, slices and time.Time all contain pointers. If you store a lot of these in memory it may be necessary to take some steps.
+
+When I've had issues with large heaps the major causes have been the following.
+
+- Lots of strings
+- Timestamps on objects using time.Time
+- Maps with slice values
+- Maps with string keys
+
+There's a lot to say about different strategies to deal with each of these. In this post I'll just talk about one idea for dealing with strings.
+
+## Strings
+[What is a string](https://syslog.ravelin.com/byte-vs-string-in-go-d645b67ca7ff)? Well, there are two parts to it. There's the string header, which tells you how long it is, and where the underlying data is. And then there's the underlying data, which is just a sequence of bytes. 
+
+When you pass a string variable to a function it is the string header that gets written to the stack, and if you keep a slice of strings, it is the string headers that appear in the slice.
+
+The string header is described by [`reflect.StringHeader`](https://golang.org/pkg/reflect/#StringHeader), which looks like the following.
+
+```go
+type StringHeader struct {
+	Data uintptr
+	Len  int
+}
+```
+
+String headers contain pointers, so we want to avoid storing strings!
+
+- If your string takes only a few fixed values then consider using integer constants instead
+- If you are storing dates and times as strings, then perhaps parse them and store the date or time as an integer
+- If you fundamentally need to keep hold of a lot of strings then read on...
+
+Let's say we're storing a hundred million strings. For simplicity, lets assume this is a single huge global `var mystrings []string`.
+
+What do we have here? Underlying `mystrings` is a `reflect.SliceHeader`, which looks similar to the `reflect.StringHeader` we've just seen.
+
+```go
+type SliceHeader struct {
+	Data uintptr
+	Len  int
+	Cap  int
+}
+```
+
+For `mystrings` Len and Cap will each be 100,000,000, and Data will point to a contiguous piece of memory large enough to contain 100,000,000 `StringHeader`s. That piece of memory contains pointers and hence will be scanned by the GC.
```

Comment:
Here too I find it helpful to mark the struct fields, so `Len` instead of Len and `Data` instead of Data.